predicators/datasets/vlm_input_data.py

```diff
@@ -1,17 +1,245 @@
+from typing import Dict, List, Optional, Sequence, Set, Tuple
 
 import numpy as np
+import PIL.Image
 
 from predicators.envs import BaseEnv
+from predicators.envs.vlm_envs import VLMPredicateEnv
+from predicators.pretrained_model_interface import GoogleGeminiVLM, \
+    VisionLanguageModel
 from predicators.settings import CFG
 from predicators.structs import Action, Dataset, GroundAtom, \
-    LowLevelTrajectory, Object, ParameterizedOption, Predicate, State, Task, \
-    _Option
+    ImageOptionTrajectory, LowLevelTrajectory, Object, ParameterizedOption, \
+    Predicate, State, Task, _Option
+
+
+def _generate_prompt_for_atom_proposals(
+        traj: ImageOptionTrajectory, trajectory_subsample_freq: int
+) -> List[Tuple[str, List[PIL.Image.Image]]]:
+    """Prompt for generating proposals for atoms.
+
+    Note that all our prompts are saved as separate txt files under the
+    'vlm_input_data_prompts/atom_proposals' folder.
+    """
+    ret_list = []
+    filepath_prefix = \
+        "predicators/datasets/vlm_input_data_prompts/atom_proposal/"
+    try:
+        with open(filepath_prefix +
+                  CFG.grammar_search_vlm_atom_proposal_prompt_type + ".txt",
+                  "r",
+                  encoding="utf-8") as f:
+            prompt = f.read()
+    except FileNotFoundError:
+        raise ValueError("Unknown VLM prompting option " +
+                         f"{CFG.grammar_search_vlm_atom_proposal_prompt_type}")
+    prompt = prompt.format(objs=[
+        str(obj.name) for obj in traj.objects if obj.name != 'dummy_goal_obj'
+    ])
+
+    if CFG.grammar_search_vlm_atom_proposal_prompt_type == "naive_each_step":
+        i = 0
+        while i < len(traj.imgs):
+            ret_list.append((prompt, traj.imgs[i]))
+            i += trajectory_subsample_freq
+    elif CFG.grammar_search_vlm_atom_proposal_prompt_type == \
+        "naive_whole_traj":
+        # NOTE: we rip out just one img from each of the state images.
+        # This is fine/works for the case where we only have one
+        # camera view, but probably will need to be amended in the future!
+        ret_list.append(
+            (prompt, [traj.imgs[i][0] for i in range(len(traj.imgs))]))
+    elif CFG.grammar_search_vlm_atom_proposal_prompt_type == \
+        "options_labels_whole_traj":
+        prompt += "\n".join(act.name + str(act.objects)
+                            for act in traj.actions)
+        # NOTE: we rip out just one img from each of the state images.
+        # This is fine/works for the case where we only have one camera
+        # view, but probably will need to be amended in the future!
+        ret_list.append(
+            (prompt, [traj.imgs[i][0] for i in range(len(traj.imgs))]))
+
+    return ret_list
+
+
+def _generate_prompt_for_scene_labelling(
+        traj: ImageOptionTrajectory,
+        atoms_list: List[str]) -> List[Tuple[str, List[PIL.Image.Image]]]:
+    """Prompt for generating labels for some set of atoms.
```

Comment:
clarify that we're trying to label the last time step of the given trajectory (right?)